File: dataforge-storage/src/main/java/hep/dataforge/storage/filestorage/FileStorageFactory.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.storage.filestorage;

import hep.dataforge.context.Context;
import hep.dataforge.meta.Meta;
import hep.dataforge.meta.MetaBuilder;
import hep.dataforge.storage.api.Storage;
import hep.dataforge.storage.api.StorageType;
import hep.dataforge.storage.commons.StorageManager;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.net.URI;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * @author Alexander Nozik
 */
public class FileStorageFactory implements StorageType {

    public static MetaBuilder buildStorageMeta(URI path, boolean readOnly, boolean monitor) {
        return new MetaBuilder("storage")
                .setValue("path", path.toString())
                .setValue("type", "file")
                .setValue("readOnly", readOnly)
                .setValue("monitor", monitor);
    }

    public static MetaBuilder buildStorageMeta(File file, boolean readOnly, boolean monitor) {
        return buildStorageMeta(file.toURI(), readOnly, monitor);
    }

    /**
     * Build local storage with Global context. Used for tests.
     *
     * @param file
     * @return
     */
    public static FileStorage buildLocal(Context context, File file, boolean readOnly, boolean monitor) {
        StorageManager manager = context.loadFeature("hep.dataforge:storage", StorageManager.class);
        return (FileStorage) manager.buildStorage(buildStorageMeta(file.toURI(),readOnly,monitor));
    }

    @Override
    public String type() {
        return "file";
    }

    @NotNull
    @Override
    public Storage build(Context context, Meta meta) {
        Path path = meta.optString("path").map(URI::create).map(Paths::get).orElse(context.getIo().getWorkDir());
        return new FileStorage(context, meta, path);
    }

}

File: dataforge-storage/src/main/kotlin/hep/dataforge/storage/filestorage/FileStorageFactory.kt
/*
 * Copyright  2018 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.storage.filestorage

import hep.dataforge.context.Context
import hep.dataforge.meta.Meta
import hep.dataforge.meta.MetaBuilder
import hep.dataforge.storage.api.Storage
import hep.dataforge.storage.api.StorageType
import hep.dataforge.storage.commons.StorageManager

import java.io.File
import java.net.URI
import java.nio.file.Path
import java.nio.file.Paths

/**
 * @author Alexander Nozik
 */
class FileStorageFactory : StorageType {

    override fun type(): String {
        return "file"
    }

    override fun build(context: Context, meta: Meta): Storage {
        val path = meta.optString("path").map<URI> { URI.create(it) }.map<Path> { Paths.get(it) }.orElse(context.io.workDir)
        return FileStorage(context, meta, path)
    }

    companion object {

        fun buildStorageMeta(path: URI, readOnly: Boolean, monitor: Boolean): MetaBuilder {
            return MetaBuilder("storage")
                    .setValue("path", path.toString())
                    .setValue("type", "file")
                    .setValue("readOnly", readOnly)
                    .setValue("monitor", monitor)
        }

        fun buildStorageMeta(file: File, readOnly: Boolean, monitor: Boolean): MetaBuilder {
            return buildStorageMeta(file.toURI(), readOnly, monitor)
        }

        /**
         * Build local storage with Global context. Used for tests.
         *
         * @param file
         * @return
         */
        fun buildLocal(context: Context, file: File, readOnly: Boolean, monitor: Boolean): FileStorage {
            val manager = context.load(StorageManager::class.java, Meta.empty())
            return manager.buildStorage(buildStorageMeta(file.toURI(), readOnly, monitor)) as FileStorage
        }
    }

}
